### PR TITLE
Feature 21 - Enemy AI

### LINKED ISSUE #21: [feature] EnemyAISystem: resting→chasing
- **Що це:** AI ворога: спочатку спокійний, потім переслідує коли бачить гравця.
- **Для чого:** базова ігрова напруга і “живий” світ.
- **Очікувані класи:** `Application::Systems::EnemyAISystem`, доменні сервіси `LOS/Pathfinding`.
- **Мінімально потрібно:**
  - якщо `Resting` і є LOS → `Chasing`;
  - якщо `Chasing` і поруч → атака;
  - інакше крок до гравця (`Pathfinding`);
  - якщо крок неможливий/втрата цілі → назад у `Resting`.
- **Edge cases:** ворог впирається в стіну; інший ворог блокує клітинку; гравця немає; мапа мала/дивна.
- **“Готово”:** тест на маленькій мапі перевіряє перехід станів і рух.

### PR DESCRIPTION
**Pull request by axumie**
## Description
Implemented EnemyAISystem, added a test and called it in the game loop. 
+fixed LOS logic and test for corner case. 

## Task ID in project
#21 

## Checklist
- [+] I went to the project and linked this PR to the task.
- [+] I set assignee and reporter on the task.
- [+] I went to the team chat and asked for review.

### CODE REVIEW

**Comment by Sage-Cat** [comment]
LMAO, шо за танці
image

File: src/app/loop/GameLoop.cpp
Comment (Sage-Cat) [inline src/app/loop/GameLoop.cpp]: Тут така фішка, що після running = false, ти вже немаєш нічого робити а в тебе після цього всеодно ще відпрацьовує Enemy AI. Зроби так, щоб у цьому випадку у тебе воно більше не працювало

File: src/domain/core/Position.hpp [hunk 1]
```diff
@@ -2,10 +2,16 @@
...
 
 namespace Domain::Core {
+
 struct Position {
     std::uint16_t x{};
     std::uint16_t y{};
-
-    friend constexpr bool operator==(Position, Position) noexcept = default;
 };
+inline bool operator==(const Position &a, const Position &b) noexcept
```
Comment (Sage-Cat) [inline src/domain/core/Position.hpp]: Нащо? Поверни назад

File: tests/app/test_enemy_ai.cpp [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+#include "domain/entities/actors/Player.hpp"
+
+using Domain::Core::Position;
+using Domain::Entities::Enemy;
+using Domain::Entities::EnemyState;
+using Domain::Entities::Player;
+
+int main()
+{
+    Domain::Core::GameState state{};
```
Comment (Sage-Cat) [inline tests/app/test_enemy_ai.cpp:17]: Додай перевірочку, що якщо користувач нажав кнопочку виходу, то у тебе EnemyAI не має нічого робити

File: src/app/systems/EnemyAISystem.cpp [hunk 3]
```diff
@@ -1,9 +1,120 @@
...
+    return p;
+}
+
+static bool occupied_by_enemy(const Domain::Core::GameState &state, Domain::Core::Position p,
+                              const Domain::Entities::Enemy *self)
+{
+    for (const auto &actor : state.actors) {
+        if (actor.get() == self)
+            continue;
+        auto *other_enemy = dynamic_cast<Domain::Entities::Enemy *>(actor.get());
```
Comment (Sage-Cat) [inline src/app/systems/EnemyAISystem.cpp]:
```suggestion
        const auto *other_enemy = dynamic_cast<const Domain::Entities::Enemy *>(actor.get());
```

File: src/app/systems/EnemyAISystem.cpp [hunk 4]
```diff
@@ -1,9 +1,120 @@
...
+        if (!other_enemy)
+            continue;
+        if (other_enemy->pos == p)
+            return true;
+    }
+    return false;
+}
+
+void EnemyAISystem::action(Domain::Core::GameState &state)
```
Comment (Sage-Cat) [inline src/app/systems/EnemyAISystem.cpp:50]: У тебе немає логіки на "втрату цілі". Якщо немає LOS, то треба щоб ворог переходив Chasing -> Resting

Аналогічно на це тести треба теж зробити

**Review by Sage-Cat — CHANGES_REQUESTED** [review]

File: tests/app/test_enemy_ai.cpp (hunk 2 above)
Comment (axumie) [reply tests/app/test_enemy_ai.cpp:17]: done

**Comment by Sage-Cat** [comment]
image

У тебе бага у LOS. Коли у тебе такий "кутик", то воно проглядає через нього. 

Що робити?

1. Новий коміт у якому розширюєш юніт тести для LOS на цей конкретний випадок з кутиком, дивишся щоб тест упав.
2. Фіксиш (моя ідея, це щоб якщо у нас ситуація такого кута, то перевіряти на прозорість обидві сусідні діагональні клітинки, якщо вони непрозорі, то видимості немає)
3. Перевіряєш щоб тест успішно виконався
4. Перевіряєш вже все разом при запуску, якщо все ок звеш на рев'ю.

File: src/app/loop/GameLoop.cpp [hunk 5]
```diff
@@ -169,10 +166,23 @@ int GameLoop::run(Domain::Core::GameState &state)
                     ++state.turn;
                 } else if constexpr (std::is_same_v<T, CmdQuit>) {
                     LOG(INFO) << "Quit requested";
-                    running = false;
+                    state.quit = true;
```
Comment (Sage-Cat) [inline src/app/loop/GameLoop.cpp:169]: Суть не змінилася, у тебе після цього коду відразу йде АІ код, тобто навіть коли у тебе користувач нажав "вихід" ворог всеодно все ще зробить один крок

**Review by Sage-Cat — CHANGES_REQUESTED** [review]

**Comment by Sage-Cat** [comment]
@codex review

File: src/app/loop/GameLoop.cpp (hunk 5 above)
Comment (axumie) [reply src/app/loop/GameLoop.cpp:169]: done

**Review by Sage-Cat — APPROVED** [review]
